x-pack/plugins/security_solution/server/assistant/tools/defend_insights/index.ts

```diff
@@ -0,0 +1,113 @@
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { PromptTemplate } from '@langchain/core/prompts';
+import { DynamicTool } from '@langchain/core/tools';
+import { LLMChain } from 'langchain/chains';
+import { OutputFixingParser } from 'langchain/output_parsers';
+
+import type { AssistantTool, AssistantToolParams } from '@kbn/elastic-assistant-plugin/server';
+import type { DefendInsightType } from '@kbn/elastic-assistant-common';
+
+import { requestHasRequiredAnonymizationParams } from '@kbn/elastic-assistant-plugin/server/lib/langchain/helpers';
+
+import { APP_UI_ID } from '../../../../common';
+import { getAnonymizedEvents } from './get_events';
+import { getDefendInsightsOutputParser } from './output_parsers';
+import { getDefendInsightsPrompt } from './prompts';
+
+export const DEFEND_INSIGHTS_TOOL_DESCRIPTION = 'Call this for Elastic Defend insights.';
+
+export interface DefendInsightsToolParams extends AssistantToolParams {
+  endpointIds: string[];
+  insightType: DefendInsightType;
+}
+
+/**
+ * Returns a tool for generating Elastic Defend configuration insights
+ */
+export const DEFEND_INSIGHTS_TOOL: AssistantTool = {
+  id: 'defend-insights',
+  name: 'defendInsightsTool',
+  description: DEFEND_INSIGHTS_TOOL_DESCRIPTION,
+  sourceRegister: APP_UI_ID,
+
+  isSupported: (params: AssistantToolParams) => {
+    const { llm, request } = params;
+
+    return requestHasRequiredAnonymizationParams(request) && llm != null;
+  },
+
+  getTool(params: AssistantToolParams) {
+    if (!this.isSupported(params)) return null;
+
+    const {
+      endpointIds,
+      insightType,
+      anonymizationFields,
+      esClient,
+      langChainTimeout,
+      llm,
+      onNewReplacements,
+      replacements,
+    } = params as DefendInsightsToolParams;
+
+    return new DynamicTool({
+      name: 'DefendInsightsTool',
+      description: DEFEND_INSIGHTS_TOOL_DESCRIPTION,
+      func: async () => {
+        if (llm == null) {
+          throw new Error('LLM is required for Defend Insights');
+        }
+
+        const anonymizedEvents = await getAnonymizedEvents({
+          endpointIds,
+          type: insightType,
+          anonymizationFields,
+          esClient,
+          onNewReplacements,
+          replacements,
+        });
+
+        const eventsContextCount = anonymizedEvents.length;
+        if (eventsContextCount === 0) {
+          return JSON.stringify({ eventsContextCount, insights: [] }, null, 2);
+        }
+
+        const outputParser = getDefendInsightsOutputParser({ type: insightType });
+        const outputFixingParser = OutputFixingParser.fromLLM(llm, outputParser);
+
+        const prompt = new PromptTemplate({
+          template: `Answer the user's question as best you can:\n{format_instructions}\n{query}`,
```

Comment:
I know its the server side, but should this be `i18n`? 

Also - remember that if displaying this in the UI, it likely is not displaying the way you want it (ex. `\n` will not be shown as a new line )

Reply:
Users won't see this, this is only fed into the LLM.